Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm

```diff
@@ -1319,12 +1417,166 @@ static WKWebExtensionError toAPI(WebExtension::Error error)
         return nil;
     }
 
-    // Cache the icon if there is only one size, usually this is a vector icon or a large bitmap.
-    if (iconDictionary.count == 1)
-        cacheLocation = result;
+    return result;
+}
+
+#if ENABLE(WK_WEB_EXTENSIONS_ICON_VARIANTS)
+static OptionSet<WebExtension::ColorScheme> toColorSchemes(id value)
+{
+    using ColorScheme = WebExtension::ColorScheme;
+
+    if (!value) {
+        // A nil value counts as all color schemes.
+        return { ColorScheme::Light, ColorScheme::Dark };
+    }
+
+    OptionSet<ColorScheme> result;
+
+    auto *array = dynamic_objc_cast<NSArray>(value);
+    if ([array containsObject:lightManifestKey])
+        result.add(ColorScheme::Light);
+
+    if ([array containsObject:darkManifestKey])
+        result.add(ColorScheme::Dark);
+
+    return result;
+}
+
+NSDictionary *WebExtension::iconsDictionaryForBestIconVariant(NSArray *variants, size_t idealPixelSize, ColorScheme idealColorScheme)
+{
+    if (!variants.count)
+        return nil;
+
+    if (variants.count == 1)
+        return variants.firstObject;
+
+    NSDictionary *bestVariant;
+    NSDictionary *fallbackVariant;
+    bool foundIdealFallbackVariant = false;
+
+    size_t bestSize = 0;
+    size_t fallbackSize = 0;
+
+    // Pick the first variant matching color scheme and/or size.
+    for (NSDictionary *variant in variants) {
+        auto *dictionary = dynamic_objc_cast<NSDictionary>(variant);
+        if (!dictionary)
+            continue;
```

Comment:
Is this worth logging some sort of error? This would be an issue with the manifest right?

Reply:
Yeah I can add some error logging.

Reply:
The error logging was handled else where already.